components/webgpu/canvas_context.rs

```diff
@@ -0,0 +1,747 @@
+            let mut snapshot =
+                Snapshot::from_vec(mapped.image_size, format, alpha_mode, result_unpadded);
+            if mapped.is_opaque {
+                snapshot.transform(SnapshotAlphaMode::Opaque, snapshot.format())
+            }
+            Some(snapshot)
+        } else {
+            None
+        }
+    }
+}
+
+impl Drop for StagingBuffer {
+    fn drop(&mut self) {
+        match self.state {
+            StagingBufferState::Unassigned => {},
+            StagingBufferState::Available(_) |
+            StagingBufferState::Mapping(_) |
+            StagingBufferState::Mapped(_) => {
+                self.global.buffer_drop(self.buffer_id);
+            },
+        }
+    }
+}
+
+/// Presentation id encodes current configuration and current image
+/// so that async presentation does not update context with older data
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
+struct PresentationId(u64);
+
+#[derive(Default)]
+pub struct WGPUExternalImages {
+    pub images: WGPUImageMap,
+    pub locked_ids: HashMap<WebGPUContextId, PresentationStagingBuffer>,
+}
+
+impl WebrenderExternalImageApi for WGPUExternalImages {
+    fn lock(&mut self, id: u64) -> (ExternalImageSource<'_>, Size2D<i32>) {
+        let id = WebGPUContextId(id);
+        let presentation = {
+            let mut webgpu_contexts = self.images.lock().unwrap();
+            webgpu_contexts
+                .get_mut(&id)
+                .and_then(|context_data| context_data.presentation.clone())
+        };
+        let Some(presentation) = presentation else {
+            return (ExternalImageSource::Invalid, Size2D::zero());
+        };
+        self.locked_ids.insert(id, presentation);
+        let presentation = self.locked_ids.get(&id).unwrap();
+        let StagingBufferState::Mapped(mapped_buffer) = &presentation.staging_buffer.state else {
+            unreachable!()
+        };
+        let size = mapped_buffer.image_size;
+        (
+            ExternalImageSource::RawData(mapped_buffer.slice()),
+            size.cast().cast_unit(),
+        )
+    }
+
+    fn unlock(&mut self, id: u64) {
+        let id = WebGPUContextId(id);
+        let Some(presentation) = self.locked_ids.remove(&id) else {
+            return;
+        };
+        let mut webgpu_contexts = self.images.lock().unwrap();
+        if let Some(context_data) = webgpu_contexts.get_mut(&id) {
+            // we use this to return staging buffer if newer exists
+            presentation.maybe_destroy(context_data);
+        } else {
+            // this will not free this adapter id in script, but that's okay
+            drop(presentation);
+        }
+    }
+}
+
+/// Staging buffer currently used for presenting the epoch.
+///
+/// Users should `put_presentation` to destroy
+#[derive(Clone)]
+pub struct PresentationStagingBuffer {
+    epoch: Epoch,
+    staging_buffer: Arc<StagingBuffer>,
+}
+
+// Only reading staging buffer
```

Comment:
It's unclear what this comment means. Maybe this:

```suggestion
// [`StagingBuffer`] only used for reading (never for writing) so it is safe to share it between threads.
```

I would also move this to right underneath the definition of `StagingBuffer`.